oioioi/su/middleware.py

```diff
@@ -39,6 +51,44 @@ def _process_request(self, request):
                 request.session[SU_UID_SESSION_KEY],
                 request.session[SU_BACKEND_SESSION_KEY],
             )
+            # Check if the user is contest admin.
+            if not request.session.get(SU_REAL_USER_IS_SUPERUSER, True):
+                # Might happen when switching to a user which then becomes a superuser.
+                if request.user.is_superuser:
+                    return redirect('su_reset')
```

Comment:
A check for whether the url is already `su_reset` is needed, otherwise this causes a redirect loop. Also it seems that `su_reset` requires POST, so probably `reset_to_real_user` should be called and a redirect returned (to the main page view ig, as a superuser might have su'ed from outside of a contest, so the contest dashboard won't be proper)..

Reply:
fixed and added a test for this